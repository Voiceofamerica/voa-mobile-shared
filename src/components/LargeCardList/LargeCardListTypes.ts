
import * as React from 'react'

import { ListItem } from '../../helpers/itemList'

export { ListItem }

export interface IconDefinition {
  icon: string | JSX.Element
  getClass?: (id: number) => string
  getStyle?: (id: number) => React.CSSProperties
  onPress: (id: number) => void
}

export interface BaseProps {
  items: ListItem[] | undefined
  onItemClick: (id: number) => void
  className?: string
  style?: React.CSSProperties
  iconButtons?: IconDefinition[]
}
