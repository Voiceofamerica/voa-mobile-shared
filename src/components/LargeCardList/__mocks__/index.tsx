
import * as React from 'react'

function LargeCardList (props) {
  return (
    React.createElement('large-card-list', { props }, props.children)
  )
}

namespace LargeCardList {
  export const Static = (props) => (
    React.createElement('static-large-card-list', { props }, props.children)
  )
}

export default LargeCardList
