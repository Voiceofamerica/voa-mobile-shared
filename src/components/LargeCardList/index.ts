
export { default } from './LargeCardList'
export * from './LargeCardList'
export * from './LargeCardListTypes'
