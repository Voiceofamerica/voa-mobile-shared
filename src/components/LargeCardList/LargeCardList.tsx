
import * as React from 'react'
import { List, ListRowProps } from 'react-virtualized'
import { Subscription } from 'rxjs/Subscription'

import { resizeObservable } from '../../helpers/windowHelper'
import { TextDirection, directionObservable } from '../../helpers/textDirectionHelper'
import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'
import LargeCard, { IconDefinition, getHeight, CARD_PADDING } from '../LargeCard'

import { BaseProps } from './LargeCardListTypes'
import { cardList, row } from './LargeCardList.scss'

export interface Props extends BaseProps, ThemeProps {
  emptyContent?: string | JSX.Element
  height?: number
  width?: number
  dir?: TextDirection
}

export interface State {
  renderHeight: number
  renderWidth: number
  cardHeight: number
  dir: TextDirection
}

const MENU_HEIGHT = 150

@themed
class LargeCardList extends React.PureComponent<Props, State> {
  state: State = {
    renderHeight: this.props.height || window.innerHeight - MENU_HEIGHT,
    renderWidth: this.props.width || window.innerWidth,
    cardHeight: getHeight(),
    dir: this.props.dir || directionObservable.getValue(),
  }

  private container: HTMLDivElement | null = null
  private resizeSub: Subscription
  private directionSub: Subscription

  componentDidMount () {
    this.resizeSub = resizeObservable.subscribe(() => {
      const renderHeight = this.getNewHeight(this.props)
      const renderWidth = this.getNewWidth(this.props)
      const cardHeight = getHeight(renderWidth)
      this.setState({ renderHeight, renderWidth, cardHeight })
    })
    this.directionSub = directionObservable.subscribe(dir => {
      if (this.props.dir) {
        return
      }

      this.setState({ dir })
    })
  }

  componentWillUnmount () {
    this.resizeSub.unsubscribe()
    this.directionSub.unsubscribe()
  }

  componentWillReceiveProps (nextProps: Props) {
    const renderHeight = this.getNewHeight(nextProps)
    const renderWidth = this.getNewWidth(nextProps)
    const cardHeight = getHeight(renderWidth)
    const dir = nextProps.dir || directionObservable.getValue()
    this.setState({ renderHeight, renderWidth, cardHeight, dir })
  }

  render () {
    const { items = [], emptyContent = '', theme = DEFAULT_THEME } = this.props
    const {
      largeCardListBackground,
    } = theme

    return (
      <div ref={this.setContainer} className={cardList} style={{ background: largeCardListBackground }}>
        {
          items.length > 0
          ? this.renderVirtualContent()
          : emptyContent
        }
      </div>
    )
  }

  private getNewHeight = ({ height = 0 }: Props) => {
    if (height > 0) {
      return height
    } else if (this.container) {
      return this.container.clientHeight
    } else {
      return window.innerHeight - MENU_HEIGHT
    }
  }

  private getNewWidth = ({ width = 0 }: Props) => {
    if (width > 0) {
      return width
    } else if (this.container) {
      return this.container.clientWidth
    } else {
      return window.innerWidth
    }
  }

  private renderVirtualContent = () => {
    const { items = [] } = this.props
    const { renderHeight, renderWidth, cardHeight } = this.state

    return (
      <List
        height={renderHeight}
        rowHeight={cardHeight + CARD_PADDING}
        rowCount={items.length}
        width={renderWidth}
        rowRenderer={this.renderRow}
      />
    )
  }

  private renderRow = ({ index, isScrolling, key, style }: ListRowProps) => {
    const { dir } = this.state
    const { items, onItemClick, iconButtons = [] } = this.props

    const { id, image, title, icon, minorText } = items![index]

    const localIconButtons = iconButtons.map(({ icon, onPress, getClass, getStyle }): IconDefinition => ({
      icon,
      getClassName: getClass ? () => getClass(id) : undefined,
      getStyle: getStyle ? () => getStyle(id) : undefined,
      onPress: () => onPress(id),
    }))

    return (
      <div key={key} className={row} style={style} dir={dir}>
        <LargeCard
          onPress={() => onItemClick(id)}
          title={title}
          minorText={minorText}
          imageUrl={image && image.hero}
          titleIcon={icon}
          iconButtons={localIconButtons}
        />
      </div>
    )
  }

  private setContainer = (el: HTMLDivElement | null) => {
    this.container = el
    const renderHeight = this.getNewHeight(this.props)
    const renderWidth = this.getNewWidth(this.props)
    const cardHeight = getHeight(renderWidth)
    this.setState({ renderHeight, renderWidth, cardHeight })
  }
}

export default LargeCardList
