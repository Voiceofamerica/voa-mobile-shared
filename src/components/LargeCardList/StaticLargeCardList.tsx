
import * as React from 'react'

import LargeCard, { IconDefinition } from '../LargeCard'

import { themed, ThemeProps, DEFAULT_THEME } from '../ThemeProvider'
import { BaseProps } from './LargeCardListTypes'
import { cardList, row, fixed } from './LargeCardList.scss'

export interface Props extends BaseProps, ThemeProps {
}

@themed
class StaticCardList extends React.Component<Props> {
  render () {
    const { items = [], theme = DEFAULT_THEME } = this.props
    const {
      largeCardListBackground,
    } = theme

    return (
      <div className={cardList} style={{ background: largeCardListBackground }}>
        {
          items.length > 0
          ? this.renderContent()
          : ''
        }
      </div>
    )
  }

  private renderContent = () => {
    const { items = [], onItemClick } = this.props

    return items.map(({ id, image, title, icon }) => (
      <div key={id} className={`${fixed} ${row}`}>
        <LargeCard
          onPress={() => onItemClick(id)}
          title={title}
          imageUrl={image && image.hero}
          titleIcon={icon}
          iconButtons={this.makeIconButtons(id)}
        />
      </div>
    ))
  }

  private makeIconButtons = (id: number): IconDefinition[] => {
    const { iconButtons = [] } = this.props
    return iconButtons.map(({ icon, onPress, getClass, getStyle }) => ({
      icon,
      getClassName: getClass ? () => getClass(id) : undefined,
      getStyle: getStyle ? () => getStyle(id) : undefined,
      onPress: () => onPress(id),
    }))
  }
}

export default StaticCardList
