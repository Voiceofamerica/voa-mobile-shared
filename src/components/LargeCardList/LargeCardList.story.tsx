
import * as React from 'react'
import { storiesOf } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import { sleep } from '../../helpers/timingHelper'
import StaticLargeCardList from './StaticLargeCardList'
import LargeCardList from './LargeCardList'
import { ListItem, IconDefinition } from './LargeCardListTypes'

const testItem: ListItem = {
  id: 1,
  image: {
    tiny: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w125_r1.png',
    thumb: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w250_r1.png',
    hero: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w500_r1.png',
  },
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const getIcon = (idx: number) => {
  idx = idx % 4
  if (idx === 0) {
    return 'video'
  } else if (idx === 1) {
    return 'audio'
  } else if (idx === 2) {
    return 'photoGallery'
  } else {
    return undefined
  }
}

const createList = (count: number) => {
  const output: ListItem[] = []
  for (let i = 0; i < count; i++) {
    output.push({
      ...testItem,
      id: i,
      icon: getIcon(i),
    })
  }
  return output
}

const list = createList(1000)
const shortList = createList(30)

const containerStyle: React.CSSProperties = {
  height: '100vh',
  display: 'flex',
  flexDirection: 'column',
}

const favorites = new Map<number, boolean>()

const icons: IconDefinition[] = [
  {
    icon: 'share',
    onPress: action('share iconButton.onPress'),
  },
  {
    icon: 'favorite',
    getStyle: (id) => {
      if (favorites.get(id)) {
        return { color: '#0061B0' }
      } else {
        return {}
      }
    },
    onPress: async (id) => {
      const originalValue = favorites.get(id)
      favorites.set(id, false)
      action('favorite iconButton pressed')(id)
      await sleep(1500)
      favorites.set(id, !originalValue)
      action('favirite iconButton done')(id)
    },
  },
  {
    icon: 'download',
    onPress: async (id) => {
      action('download iconButton pressed')(id)
      await sleep(1500)
      action('download iconButton failed')(id)
      throw new Error()
    },
  },
]

storiesOf(LargeCardList.name, module)
  .add('display', () => (
    <div style={containerStyle}>
      <style>{'body { margin: 0; }'}</style>
      <LargeCardList
        items={list}
        onItemClick={action('onItemClick')}
        iconButtons={icons}
      />
    </div>
  ))
  .add('rtl display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <LargeCardList
        items={list}
        onItemClick={action('onItemClick')}
        dir='rtl'
        iconButtons={icons}
      />
    </div>
  ))
  .add('static height display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <LargeCardList
        items={list}
        onItemClick={action('onItemClick')}
        height={200}
        iconButtons={icons}
      />
    </div>
  ))
  .add('contained height display', () => (
    <div style={{height: 300, display: 'flex'}}>
      <style>{'body { margin: 0; }'}</style>
      <LargeCardList
        items={list}
        onItemClick={action('onItemClick')}
        iconButtons={icons}
      />
    </div>
  ))

storiesOf(`${StaticLargeCardList.name}.Static`, module)
  .add('display', () => (
    <div>
      <style>{'body { margin: 0; }'}</style>
      <StaticLargeCardList
        items={shortList}
        onItemClick={action('onItemClick')}
        iconButtons={icons}
      />
    </div>
  ))
  .add('rtl display', () => (
    <div dir='rtl'>
      <style>{'body { margin: 0; }'}</style>
      <StaticLargeCardList
        items={shortList}
        onItemClick={action('onItemClick')}
        iconButtons={icons}
      />
    </div>
  ))
