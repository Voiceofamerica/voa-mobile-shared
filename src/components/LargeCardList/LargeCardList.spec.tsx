
import * as React from 'react'
import { create } from 'react-test-renderer'
import { shallow } from 'enzyme'

import { ListItem } from './LargeCardListTypes'
import LargeCardList, { Props, State } from './LargeCardList'

jest.mock('react-virtualized')
jest.mock('../LargeCard')
jest.mock('../SvgIcon')

const testItem: ListItem = {
  id: 1,
  image: {
    hero: 'https://gdb.voanews.com/3E1C440D-1C78-4B88-84F7-ECB12FB5B874_cx3_cy0_cw97_w650_r1.png',
  },
  minorText: '15:08',
  title: 'President Trump: From North Korea to Iran, we confront dangerous governments',
}

const createList = (count: number) => {
  const output: ListItem[] = []
  for (let i = 0; i < count; i++) {
    output.push({
      ...testItem,
      id: i,
    })
  }
  return output
}

const list = createList(1000)

describe(`<${LargeCardList.name} />`, () => {
  describe('shapshots', () => {
    it('should render only the necessary items', () => {
      const element = create((
        <LargeCardList items={list} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should render the empty content if no items', () => {
      const element = create((
        <LargeCardList items={[]} onItemClick={jest.fn()} emptyContent='I am empty' />
      ))

      expect(element).toMatchSnapshot()
    })
    it('should not throw if items is undefined', () => {
      const element = create((
        <LargeCardList items={undefined} onItemClick={jest.fn()} />
      ))

      expect(element).toMatchSnapshot()
    })
  })
})
