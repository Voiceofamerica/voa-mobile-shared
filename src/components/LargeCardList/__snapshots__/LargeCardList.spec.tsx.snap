// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LargeCardList /> shapshots should not throw if items is undefined 1`] = `
<div
  className="cardList"
  style={
    Object {
      "background": "#B4BCC2",
    }
  }
>
  
</div>
`;

exports[`<LargeCardList /> shapshots should render only the necessary items 1`] = `
<div
  className="cardList"
  style={
    Object {
      "background": "#B4BCC2",
    }
  }
>
  <react-virtualized-list
    height={618}
    rowCount={1000}
    rowHeight={NaN}
    rowRenderer={[Function]}
    width={1024}
  >
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
    <div
      className="row"
      dir="ltr"
      style={Object {}}
    />
  </react-virtualized-list>
</div>
`;

exports[`<LargeCardList /> shapshots should render the empty content if no items 1`] = `
<div
  className="cardList"
  style={
    Object {
      "background": "#B4BCC2",
    }
  }
>
  I am empty
</div>
`;
